File: eventify-platform/services/api-gateway/src/metrics.js
const client = require('prom-client');

// Collect default metrics (CPU, memory, etc.)
const collectDefaultMetrics = client.collectDefaultMetrics;
collectDefaultMetrics({ timeout: 5000 });

// ===== 1. API-LEVEL METRICS =====

// Request count counter
const httpRequestCounter = new client.Counter({
  name: 'http_requests_total',
  help: 'Total number of HTTP requests',
  labelNames: ['method', 'route', 'status_code']
});

// Request latency histogram (p50, p95, p99)
const httpRequestDuration = new client.Histogram({
  name: 'http_request_duration_seconds',
  help: 'HTTP request latency in seconds',
  labelNames: ['method', 'route', 'status_code'],
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
});

// Active requests gauge (concurrency)
const httpActiveRequests = new client.Gauge({
  name: 'http_active_requests',
  help: 'Number of active HTTP requests',
  labelNames: ['method']
});

// Error tracking counter
const httpErrorsTotal = new client.Counter({
  name: 'http_errors_total',
  help: 'Total number of HTTP errors',
  labelNames: ['method', 'route', 'status_code', 'error_type']
});

// ===== 2. SERVICE-LEVEL INSTRUMENTATION =====

// JWT authentication operations
const jwtOperationDuration = new client.Histogram({
  name: 'jwt_operation_duration_seconds',
  help: 'JWT operation execution time',
  labelNames: ['operation', 'status'],
  buckets: [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5]
});

const jwtOperationTotal = new client.Counter({
  name: 'jwt_operations_total',
  help: 'Total JWT operations',
  labelNames: ['operation', 'status']
});

// Rate limiting operations
const rateLimitHits = new client.Counter({
  name: 'rate_limit_hits_total',
  help: 'Total rate limit hits',
  labelNames: ['action']
});

// ===== 3. DEPENDENCY OBSERVABILITY =====

// Downstream service call metrics
const downstreamRequestDuration = new client.Histogram({
  name: 'downstream_request_duration_seconds',
  help: 'Downstream service request latency',
  labelNames: ['service', 'status'],
  buckets: [0.01, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
});

const downstreamRequestErrors = new client.Counter({
  name: 'downstream_request_errors_total',
  help: 'Total downstream service request errors',
  labelNames: ['service', 'error_type']
});

// Redis cache operations
const cacheOperationDuration = new client.Histogram({
  name: 'cache_operation_duration_seconds',
  help: 'Cache operation latency',
  labelNames: ['operation', 'status'],
  buckets: [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5]
});

const cacheOperationErrors = new client.Counter({
  name: 'cache_operation_errors_total',
  help: 'Total cache operation errors',
  labelNames: ['operation', 'error_type']
});

// ===== MIDDLEWARE FUNCTIONS =====

/**
 * Middleware to track request metrics
 */
function metricsMiddleware(req, res, next) {
  const start = Date.now();
  
  // Increment active requests
  httpActiveRequests.inc({ method: req.method });
  
  // Track response
  res.on('finish', () => {
    const duration = (Date.now() - start) / 1000;
    const route = req.route ? req.route.path : req.path;
    const statusCode = res.statusCode.toString();
    
    // Record metrics
    httpRequestCounter.inc({ method: req.method, route, status_code: statusCode });
    httpRequestDuration.observe({ method: req.method, route, status_code: statusCode }, duration);
    
    // Track errors (4xx, 5xx)
    if (res.statusCode >= 400) {
      const errorType = res.statusCode >= 500 ? 'server_error' : 'client_error';
      httpErrorsTotal.inc({ method: req.method, route, status_code: statusCode, error_type: errorType });
    }
    
    // Decrement active requests
    httpActiveRequests.dec({ method: req.method });
  });
  
  next();
}

/**
 * Track JWT verification operations
 */
function trackJWTOperation(operation, success) {
  const status = success ? 'success' : 'failure';
  jwtOperationTotal.inc({ operation, status });
}

/**
 * Track JWT verification with timing
 */
function trackJWTVerification(fn) {
  return (...args) => {
    const end = jwtOperationDuration.startTimer({ operation: 'verify', status: 'pending' });
    try {
      const result = fn(...args);
      end({ operation: 'verify', status: 'success' });
      trackJWTOperation('verify', true);
      return result;
    } catch (error) {
      end({ operation: 'verify', status: 'failure' });
      trackJWTOperation('verify', false);
      throw error;
    }
  };
}

/**
 * Track rate limit hits
 */
function trackRateLimit(action) {
  rateLimitHits.inc({ action });
}

module.exports = {
  client,
  metricsMiddleware,
  trackJWTOperation,
  trackJWTVerification,
  trackRateLimit,
  httpRequestCounter,
  httpRequestDuration,
  httpActiveRequests,
  httpErrorsTotal,
  jwtOperationDuration,
  jwtOperationTotal,
  rateLimitHits,
  downstreamRequestDuration,
  downstreamRequestErrors,
  cacheOperationDuration,
  cacheOperationErrors
};

File: eventify-platform/services/notification-service/metrics.py
"""
Comprehensive Prometheus metrics for Notification Service
Implements API-level, service-level, and dependency observability
"""
from prometheus_client import Counter, Histogram, Gauge, generate_latest, CONTENT_TYPE_LATEST, REGISTRY
import time
from functools import wraps

# ===== 1. API-LEVEL METRICS =====

http_requests_total = Counter(
    'http_requests_total',
    'Total number of HTTP requests',
    ['method', 'endpoint', 'status_code']
)

http_request_duration_seconds = Histogram(
    'http_request_duration_seconds',
    'HTTP request latency in seconds',
    ['method', 'endpoint', 'status_code'],
    buckets=[0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
)

http_active_requests = Gauge(
    'http_active_requests',
    'Number of active HTTP requests',
    ['method']
)

http_errors_total = Counter(
    'http_errors_total',
    'Total number of HTTP errors',
    ['method', 'endpoint', 'status_code', 'error_type']
)

# ===== 2. SERVICE-LEVEL INSTRUMENTATION =====

notification_send_duration_seconds = Histogram(
    'notification_send_duration_seconds',
    'Notification send operation execution time',
    ['notification_type', 'status'],
    buckets=[0.01, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
)

notification_send_operations_total = Counter(
    'notification_send_operations_total',
    'Total notification send operations',
    ['notification_type', 'status']
)

notifications_sent_total = Counter(
    'notifications_sent_total',
    'Total notifications sent successfully',
    ['type']
)

notifications_failed_total = Counter(
    'notifications_failed_total',
    'Total notifications that failed to send',
    ['type', 'error_type']
)

notification_queue_size = Gauge(
    'notification_queue_size',
    'Current notification queue size'
)

# ===== 3. DEPENDENCY OBSERVABILITY =====

# Redis operations
redis_operation_duration_seconds = Histogram(
    'redis_operation_duration_seconds',
    'Redis operation latency',
    ['operation', 'status'],
    buckets=[0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5]
)

redis_operation_errors_total = Counter(
    'redis_operation_errors_total',
    'Total Redis operation errors',
    ['operation', 'error_type']
)

# RabbitMQ operations
rabbitmq_consume_duration_seconds = Histogram(
    'rabbitmq_consume_duration_seconds',
    'RabbitMQ message consume latency',
    ['routing_key', 'status'],
    buckets=[0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5]
)

rabbitmq_consume_errors_total = Counter(
    'rabbitmq_consume_errors_total',
    'Total RabbitMQ consume errors',
    ['routing_key', 'error_type']
)

rabbitmq_messages_processed_total = Counter(
    'rabbitmq_messages_processed_total',
    'Total RabbitMQ messages processed',
    ['routing_key', 'status']
)

# External API calls (User Service)
external_api_call_duration_seconds = Histogram(
    'external_api_call_duration_seconds',
    'External API call latency',
    ['service', 'operation', 'status'],
    buckets=[0.01, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
)

external_api_call_errors_total = Counter(
    'external_api_call_errors_total',
    'Total external API call errors',
    ['service', 'operation', 'error_type']
)

# SMTP operations
smtp_send_duration_seconds = Histogram(
    'smtp_send_duration_seconds',
    'SMTP email send latency',
    ['status'],
    buckets=[0.1, 0.5, 1, 2.5, 5, 10, 30]
)

smtp_send_errors_total = Counter(
    'smtp_send_errors_total',
    'Total SMTP send errors',
    ['error_type']
)

# ===== HELPER FUNCTIONS =====

def track_request(method, endpoint, status_code, duration):
    """Track HTTP request metrics"""
    http_requests_total.labels(method=method, endpoint=endpoint, status_code=status_code).inc()
    http_request_duration_seconds.labels(method=method, endpoint=endpoint, status_code=status_code).observe(duration)
    
    if status_code >= 400:
        error_type = 'server_error' if status_code >= 500 else 'client_error'
        http_errors_total.labels(method=method, endpoint=endpoint, status_code=status_code, error_type=error_type).inc()

def track_notification_send(notification_type):
    """Decorator to track notification send operations"""
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            start = time.time()
            status = 'failure'
            try:
                result = func(*args, **kwargs)
                status = 'success'
                notifications_sent_total.labels(type=notification_type).inc()
                return result
            except Exception as e:
                status = 'failure'
                error_type = type(e).__name__
                notifications_failed_total.labels(type=notification_type, error_type=error_type).inc()
                raise
            finally:
                duration = time.time() - start
                notification_send_duration_seconds.labels(notification_type=notification_type, status=status).observe(duration)
                notification_send_operations_total.labels(notification_type=notification_type, status=status).inc()
        return wrapper
    return decorator

def track_redis_operation(operation):
    """Track Redis operation with timing"""
    class RedisTimer:
        def __init__(self, operation):
            self.operation = operation
            self.start = time.time()
            
        def __enter__(self):
            return self
            
        def __exit__(self, exc_type, exc_val, exc_tb):
            duration = time.time() - self.start
            if exc_type is None:
                status = 'success'
                redis_operation_duration_seconds.labels(operation=self.operation, status=status).observe(duration)
            else:
                status = 'failure'
                error_type = exc_type.__name__ if exc_type else 'Unknown'
                redis_operation_duration_seconds.labels(operation=self.operation, status=status).observe(duration)
                redis_operation_errors_total.labels(operation=self.operation, error_type=error_type).inc()
            return False
    
    return RedisTimer(operation)

def track_rabbitmq_consume(routing_key):
    """Track RabbitMQ message consume operation"""
    class RabbitMQTimer:
        def __init__(self, routing_key):
            self.routing_key = routing_key
            self.start = time.time()
            
        def __enter__(self):
            return self
            
        def __exit__(self, exc_type, exc_val, exc_tb):
            duration = time.time() - self.start
            if exc_type is None:
                status = 'success'
                rabbitmq_consume_duration_seconds.labels(routing_key=self.routing_key, status=status).observe(duration)
                rabbitmq_messages_processed_total.labels(routing_key=self.routing_key, status=status).inc()
            else:
                status = 'failure'
                error_type = exc_type.__name__ if exc_type else 'Unknown'
                rabbitmq_consume_duration_seconds.labels(routing_key=self.routing_key, status=status).observe(duration)
                rabbitmq_consume_errors_total.labels(routing_key=self.routing_key, error_type=error_type).inc()
                rabbitmq_messages_processed_total.labels(routing_key=self.routing_key, status=status).inc()
            return False
    
    return RabbitMQTimer(routing_key)

def track_external_api_call(service, operation):
    """Track external API call with timing"""
    class APITimer:
        def __init__(self, service, operation):
            self.service = service
            self.operation = operation
            self.start = time.time()
            
        def __enter__(self):
            return self
            
        def __exit__(self, exc_type, exc_val, exc_tb):
            duration = time.time() - self.start
            if exc_type is None:
                status = 'success'
                external_api_call_duration_seconds.labels(service=self.service, operation=self.operation, status=status).observe(duration)
            else:
                status = 'failure'
                error_type = exc_type.__name__ if exc_type else 'Unknown'
                external_api_call_duration_seconds.labels(service=self.service, operation=self.operation, status=status).observe(duration)
                external_api_call_errors_total.labels(service=self.service, operation=self.operation, error_type=error_type).inc()
            return False
    
    return APITimer(service, operation)

def track_smtp_send():
    """Track SMTP email send with timing"""
    class SMTPTimer:
        def __init__(self):
            self.start = time.time()
            
        def __enter__(self):
            return self
            
        def __exit__(self, exc_type, exc_val, exc_tb):
            duration = time.time() - self.start
            if exc_type is None:
                status = 'success'
                smtp_send_duration_seconds.labels(status=status).observe(duration)
            else:
                status = 'failure'
                error_type = exc_type.__name__ if exc_type else 'Unknown'
                smtp_send_duration_seconds.labels(status=status).observe(duration)
                smtp_send_errors_total.labels(error_type=error_type).inc()
            return False
    
    return SMTPTimer()
